piptools/resolver.py

```diff
@@ -119,9 +119,9 @@ def resolve(self, max_rounds=10):
 
     def _check_constraints(self):
         for constraint in chain(self.our_constraints, self.their_constraints):
-            if constraint.link is not None and not constraint.editable:
-                msg = ('pip-compile does not support URLs as packages, unless they are editable. '
-                       'Perhaps add -e option?')
+            if ((constraint.link and not constraint.editable
+                 and not constraint.link.is_artifact and not is_pinned_requirement(constraint))):
```

Comment:
Ah, yeah, this is something I wanted to run by you. The motivation is to restrict the input so pip-compile can take it at face value. In particular, supporting unpinned vcs links seems tough since:
* pip doesn't seem to expose the "true" version of a vcs link unless it's been cloned and installed (it just seems to assume the provided version is accurate for the `@ref` in the link)
* internal logic assumes that constraints are pinned later on (ie, removing _check_constraints raises a "expected pinned or editable..." TypeError later on)

Thoughts?